src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Collections/FlatArrayIntervalTree`1.cs

```diff
@@ -0,0 +1,430 @@
+    /// incur a delegate allocation to sort the values.  If callers can avoid that allocation by pre-sorting the values,
+    /// they should do so and call <see cref="CreateFromSorted"/> instead.
+    /// </summary>
+    /// <remarks>
+    /// <paramref name="values"/> will be sorted in place.
+    /// </remarks>
+    public static FlatArrayIntervalTree<T> CreateFromUnsorted<TIntrospector>(in TIntrospector introspector, SegmentedList<T> values)
+        where TIntrospector : struct, IIntervalIntrospector<T>
+    {
+        var localIntrospector = introspector;
+        values.Sort((t1, t2) => localIntrospector.GetSpan(t1).Start - localIntrospector.GetSpan(t2).Start);
+        return CreateFromSorted(introspector, values);
+    }
+
+    /// <summary>
+    /// Creates an interval tree from a sorted list of values.  This is more efficient than creating from an unsorted
+    /// list as building doesn't need to figure out where the nodes need to go n-log(n) and doesn't have to rebalance
+    /// anything (again, another n-log(n) operation).  Rebalancing is particularly expensive as it involves tons of
+    /// pointer chasing operations, which is both slow, and which impacts the GC which has to track all those writes.
+    /// </summary>
+    /// <remarks>
+    /// <list type="bullet">The values must be sorted such that given any two elements 'a' and 'b' in the list, if 'a'
+    /// comes before 'b' in the list, then it's "start position" (as determined by the introspector) must be less than
+    /// or equal to 'b's start position.  This is a requirement for the algorithm to work correctly.
+    /// </list>
+    /// <list type="bullet">The <paramref name="values"/> list will be mutated as part of this operation.
+    /// </list>
+    /// </remarks>
+    public static FlatArrayIntervalTree<T> CreateFromSorted<TIntrospector>(in TIntrospector introspector, SegmentedList<T> values)
+        where TIntrospector : struct, IIntervalIntrospector<T>
+    {
+#if DEBUG
+        var localIntrospector = introspector;
+        Debug.Assert(values.IsSorted(Comparer<T>.Create((t1, t2) => localIntrospector.GetSpan(t1).Start - localIntrospector.GetSpan(t2).Start)));
+#endif
+
+        if (values.Count == 0)
+            return Empty;
+
+        // Create the array to sort the binary search tree nodes in.
+        var array = new SegmentedArray<Node>(values.Count);
+
+        // Place the values into the array in a way that will create a complete binary tree.
+        BuildCompleteTreeTop(values, array);
+
+        // Next, do a pass over the entire tree, updating each node to point at the max end node in its subtree.
+        ComputeMaxEndNodes(array, 0, in introspector);
+
+        return new FlatArrayIntervalTree<T>(array);
+
+        static void BuildCompleteTreeTop(SegmentedList<T> source, SegmentedArray<Node> destination)
+        {
+            // The nature of a complete tree is that the last level always only contains the odd remaining numbers.
```

Comment:
>the odd remaining numbers [](http://example.com/codeflow?start=89&length=25)

this was a bit unclear, maybe something like "values from the odd indexes"

Reply:
yup.  will fix.

Reply:
also, these are the even indices, right?

Reply:
ugh.  yes.  computer indices. w ill update all docs in followup.

Reply:
fixed in next pr.